srtcore/core.cpp

```diff
@@ -9701,10 +9691,243 @@ int srt::CUDT::processData(CUnit* in_unit)
             return -1;
     }
 
+    return 0;
+}
+
+CUDT::time_point srt::CUDT::getPktTsbPdTime(void*, const CPacket& packet)
+{
+    return m_pRcvBuffer->getPktTsbPdTime(packet.getMsgTimeStamp());
+}
+
+SRT_ATR_UNUSED static const char *const s_rexmitstat_str[] = {"ORIGINAL", "REXMITTED", "RXS-UNKNOWN"};
+
+int srt::CUDT::handleSocketPacketReception(const vector<CUnit*>& incoming, bool& w_new_inserted, bool& w_was_sent_in_order, CUDT::loss_seqs_t& w_srt_loss_seqs)
+{
+    bool excessive SRT_ATR_UNUSED = true; // stays true unless it was successfully added
+
+    // Loop over all incoming packets that were filtered out.
+    // In case when there is no filter, there's just one packet in 'incoming',
+    // the one that came in the input of this function.
+    for (vector<CUnit *>::const_iterator unitIt = incoming.begin(); unitIt != incoming.end(); ++unitIt)
+    {
+        CUnit *  u    = *unitIt;
+        CPacket &rpkt = u->m_Packet;
+        const int pktrexmitflag = m_bPeerRexmitFlag ? (rpkt.getRexmitFlag() ? 1 : 0) : 2;
+        const bool retransmitted = pktrexmitflag == 1;
+
+        int buffer_add_result;
+        bool adding_successful = true;
+
+        // m_iRcvLastSkipAck is the base sequence number for the receiver buffer.
+        // This is the offset in the buffer; if this is negative, it means that
+        // this sequence is already in the past and the buffer is not interested.
+        // Meaning, this packet will be rejected, even if it could potentially be
+        // one of missing packets in the transmission.
+        int32_t offset = CSeqNo::seqoff(m_iRcvLastSkipAck, rpkt.m_iSeqNo);
+
+        IF_HEAVY_LOGGING(const char *exc_type = "EXPECTED");
+
+        if (offset < 0)
+        {
+            IF_HEAVY_LOGGING(exc_type = "BELATED");
+            time_point pts = getPktTsbPdTime(NULL, rpkt);
+
+            enterCS(m_StatsLock);
+            const double bltime = (double) CountIIR<uint64_t>(
+                    uint64_t(m_stats.traceBelatedTime) * 1000,
+                    count_microseconds(steady_clock::now() - pts), 0.2);
+
+            m_stats.traceBelatedTime = bltime / 1000.0;
+            m_stats.rcvr.recvdBelated.count(rpkt.getLength());
+            leaveCS(m_StatsLock);
+            HLOGC(qrlog.Debug,
+                    log << CONID() << "RECEIVED: seq=" << rpkt.m_iSeqNo << " offset=" << offset << " (BELATED/"
+                    << s_rexmitstat_str[pktrexmitflag] << ") FLAGS: " << rpkt.MessageFlagStr());
+            continue;
+        }
+
+        // This is executed only when bonding is enabled and only
+        // with the new buffer (in which case the buffer is in the group).
+
+        const int avail_bufsize = (int) getAvailRcvBufferSizeNoLock();
```

Comment:
This function doesn't match the codes removed from `processData()` below, several PR were lost.

Reply:
What exactly? I can see only changes in logs and slight improvement for the offset counting. Some parts are left up to the calling function (emergency closing is handled through returning -2). You can cut off since this line:

```
        UniqueLock recvbuf_acklock(m_RcvBufferLock);
```

and except the last two branches (which remain at the original location and are unchanged) you can see no significant changes in the `handleSocketPacketReception`.

Reply:
> What exactly?

e.g. #2465 #2467

> I can see only changes in logs and slight improvement for the offset counting...you can see no significant changes

Yes, these are only minor improvement, so you can silently revert them?

Reply:
e.g. #2465  #2467

Reply:
> I can see only changes in logs and slight improvement for the offset counting.

Your refactor is great, but I don't think we can silently revert PR.

Reply:
Hey, one moment. @maxsharabayko please join.

Since when is `m_iRcvLastAck` the base sequence number for the receiver buffer?

This was so in the old UDT, and it was sharing the number that was latest received as ACK as well as the sequence number of the first cell in the receiver buffer. This has changed since the first versions of SRT when the TLPKTDROP was introduced and so a separate field `m_iRcvLastSkipAck`, which has become the true beginning of the receiver buffer. Here is an excerpt from the `CUDT::tsbpd` function taken from 1.4.4 version, which executes the drop functionality in order to reach the after-drop packets:

```
                /* Packet ready to play according to time stamp but... */
                int seqlen = CSeqNo::seqoff(self->m_iRcvLastSkipAck, skiptoseqno);

                if (skiptoseqno != SRT_SEQNO_NONE && seqlen > 0)
                {
                    /*
                     * skiptoseqno != SRT_SEQNO_NONE,
                     * packet ready to play but preceeded by missing packets (hole).
                     */

                    self->updateForgotten(seqlen, self->m_iRcvLastSkipAck, skiptoseqno);
                    self->m_pRcvBuffer->skipData(seqlen);

                    self->m_iRcvLastSkipAck = skiptoseqno;
```

Here you can see that the call to `skipData` is getting only the number of cells to drop as the old receiver buffer had no idea about the sequence numbers at all. The code is almost the same in version 1.2.0, the very first line added to Github. The base sequence number to calculate `seqlen` is taken from `m_iRcvLastSkipAck` and this is the real base sequence number for the receiver buffer.

Reply:
So the conclusion is using `m_iRcvLastSkipAck` as the recv buffer base seq? Then I would suggest to change `getAvailRcvBufferSizeNoLock()` to `return m_pRcvBuffer->getAvailSize(m_iRcvLastSkipAck);` to keep `offset` consistent with avail buffer size.

Reply:
And `m_pRcvBuffer->strFullnessState(qrlog.Warn.CheckEnabled(), m_iRcvLastAck, steady_clock::now())` should be fixed to `m_iRcvLastSkipAck`